src/providers/ogr/qgsogrfeatureiterator.cpp

```diff
@@ -282,20 +296,38 @@ bool QgsOgrFeatureIterator::fetchFeature( QgsFeature &feature )
 
   gdal::ogr_feature_unique_ptr fet;
 
-  while ( fet.reset( OGR_L_GetNextFeature( mOgrLayer ) ), fet )
+  // OSM layers (especially large ones) need the GDALDataset::GetNextFeature() call rather than OGRLayer::GetNextFeature()
+  // see more details here: https://trac.osgeo.org/gdal/wiki/rfc66_randomlayerreadwrite
+  if ( mSource->mDriverName == QLatin1String( "OSM" ) ) 
   {
-    if ( !readFeature( std::move( fet ), feature ) )
-      continue;
 
-    if ( !mFilterRect.isNull() && ( !feature.hasGeometry() || feature.geometry().isEmpty() ) )
-      continue;
-
-    // we have a feature, end this cycle
-    feature.setValid( true );
-    geometryToDestinationCrs( feature, mTransform );
-    return true;
+    while ( fet.reset( GDALDatasetGetNextFeature( mConn->ds, nullptr, nullptr, nullptr, nullptr )), fet )
+    {
+      if ( checkFeature( fet, feature ) ) 
```

Comment:
this should check that the feature belongs to the layer of interest. It is returned in the second parameter of GDALDatasetGetNextFeature()